### PR TITLE
Solve galaxy s5 issues using Nordpol

### LINKED ISSUE #9: Samsung Galaxy S5 NFC Service Crashes
See http://forum.yubico.com/viewtopic.php?t=1646

I can read the NEO once with the S5 without problem, trying to read it once again just seams to hang the nfc processes on the phone. If i disable NFC on the phone and enable it again I'm able to read the NEO once more. When it's not possible to read the NEO it's not possible to read any other NFC tags. Same scenario works with out problem on S3,S4,Z3 and so on.

Background:
Model: SM-G900F
Android 4.4.2
Kernel 3.4.0-2089850
Baseband: G900FXXU1ANG2

Symptoms:
After one read(successful OR failed), it will not read again

Workarounds:
- wait at least a minute then try again(does not always work, 50/50...)
- disable/enable NFC
- lock/unlock the phone

Comments:
On the first attempt, the S5 seems to be able to successfully read the Neo only about 50% of the time, sometimes I get error "Unfortunately, NFC Service has stopped", but most times it just simply will not read again until doing one of the above 'workarounds'. After a failed read, if I immediately exit the Yubico Authenticator then re-load it, I get message "NFC is disabled" but when I check, it is actually enabled. So it seems that the NFC service is crashing for some reason. Note: I have never seen "NFC is disabled" message after a successful read, but it still will not read again until doing one of the workarounds above.

Additional Info:
Adb logcat says when scanning Neo first time:

D/NativeNfcTag(11676): connectWithStatus- technology =3
D/NativeNfcTag(11676): Connect to a tech with a different handle
D/NativeNfcTag(11676): Starting background presence check
D/NfcDispatcher(11676): tryStartActivity. Send intent.
V/ApplicationPolicy( 807): isApplicationStateBlocked userId 0 pkgname com.android.nfc
D/NfcService(11676): tag value : none
V/SmartFaceService - 3rd party pause( 807): onReceive [android.intent.action.ACTIVITY_STATE/com.android.nfc/create]
V/SmartFaceService - 3rd party pause( 807): onReceive [android.intent.action.ACTIVITY_STATE/com.android.nfc/pause]
D/NativeNfcTag(11676): Tag lost, restarting polling loop
D/NativeNfcTag(11676): Stopping background presence check
E/NfcNfa (11676): nfa_dm_act_deactivate (): invalid protocol, mode or state

When trying to scan ANY tag after this, nothing happens.

The 2 last lines look not to good, in my eyes at least... Doing the same on S3 doesn't provide as much detail

D/NfcDispatcher( 2940): tryStartActivity. Send intent.
D/NfcService( 2940): tag value : none
D/NativeNfcTag( 2940): Tag lost, restarting polling loop

But on S3 it works all the time.

### PR DESCRIPTION
**Pull request by Franzaine**
This pull request fixes #9 by introducing NFC handling via the Nordpol library including specific tweaks for Samsung devices. It will automatically choose the best way of handling NFC on different Android devices.

### CODE REVIEW

**Comment by dainnilsson** [comment]
Great, I'm very happy if this solves the issues with NFC! Unfortunately I
don't have access to a Samsung S5 or S6 to test with, but I'll get around
to testing this on the devices I do have access to soon.

**Comment by Franzaine** [comment]
@dainnilsson It seems like this pull request solved most of the issues with the S5 and actually makes the app usable on that device https://github.com/Yubico/yubioath-android/issues/9

Ready for you to review!

**Comment by dainnilsson** [comment]
Excellent. I have a lot of other things on my plate right now, but as soon as I get a chance I'll take a deeper look and get this out.

**Comment by justinnoah** [comment]
@dainnilsson If you are in need of a test device, I offer mine as a guinea pig. I can run things and send you debug output/etc if you'd like.

**Comment by dainnilsson** [comment]
I've tested this a bit now, on some non S5 devices. On eissue that I'm finding is that the app no longer gives you any audio feedback when tapping the YubiKey. Before it would signal both the device was "searching", and when a successful connection was made. Now it's completely silent. I haven't had a chance to look into the code, but is there anything in the Nordpol library that is causing thing and can it be changed to bring back the old behavior?

**Comment by petter-fidesmo** [comment]
@dainnilsson I looked into nordpol and it did indeed default to silencing the NFC sounds. We changed that so it now defaults to turning it on (which is also the Android default). The last commit updates the dependency to the latest version which should fix the issue. Have nice weekend!

**Comment by dainnilsson** [comment]
Thanks, I just tested this. It does add the confirmation "bading!" once the connection is made, but it doesn't make the "searching"-sound when the key is near but not yet in the correct spot. This feedback is quite useful for finding the perfect location to hold the YubiKey. Is there any way to re-activate this as well?

**Comment by Franzaine** [comment]
@dainnilsson I have never heard this second sound that you speak of but I do see its value as immediate feedback to the user.

I have tried to reproduce the sounds but have so far not succeeded on my Nexus 5. What phone do you use when you get the sound and does it work on multiple devices?
Thanks, Theo

**Comment by dainnilsson** [comment]
I've heard it on every device I've tested with (mainly Nexus devices, including 5). It's especially noticeable when you place the Yubikey slightly away from the "sweet spot". I managed to find the sounds here: https://android.googlesource.com/platform/packages/apps/Nfc/+/master/res/raw/
The one named "start.ogg" that loops while the key is near but hasn't yet made a solid connection. It ends in either "end.ogg" on success (which is the only one that I'm hearing with Nordpol) or "error.ogg" on failure.

**Comment by petter-fidesmo** [comment]
After lots of testing with a Yubikey we think we have found the cause for the searching sound!

What seems to be going on is that in the old implementation Android is initially reading this device as an NDEF device (even though the intent filter requests an ISO device). This is default behaviour on Android. For Android 4.4+ nordpol default to the new [reader mode](http://developer.android.com/reference/android/nfc/NfcAdapter.html#enableReaderMode%28android.app.Activity,%20android.nfc.NfcAdapter.ReaderCallback,%20int,%20android.os.Bundle%29). It solves a bunch of problems on several devices (such as the problem with the S5) and it provides much more fine grained control over the NFC stack. Therefore nordpol disables the NDEF check using the [following flag]( http://developer.android.com/reference/android/nfc/NfcAdapter.html#FLAG_READER_SKIP_NDEF_CHECK). This removed the unnecessary communication to find out if the card has an NDEF record as we are not interested in this.

The sound you are hearing in the old implementation is played because it finds (or fails to find) an NDEF record. Since the application has requested [foreground dispatch](http://developer.android.com/reference/android/nfc/NfcAdapter.html#enableForegroundDispatch%28android.app.Activity,%20android.app.PendingIntent,%20android.content.IntentFilter[],%20java.lang.String[][]%29), the NDEF message is discarded.

The sound is not indicating that the NFC device is closing in on the right place, but rather that an NDEF record is found/not found. Without the NDEF check, the tag is directly forwarded to the application. This should reduce the time the user have to wait for the OTP code to be generated.

**Comment by dainnilsson** [comment]
Understood. That's a shame, I'm sure others have interpreted the sound in the same way I have (and in a way it actually does give you helpful feedback, since the positioning for scanning an NDEF tag is the same). Thanks for looking into this! I'm going to merge now and hope to get a new release out pretty soon (not sure if I'll get to it today or not). I anticipate some backlash from non-S5 users. If it's too bad we'll have to do something about it.

File: src/main/com/yubico/yubioath/MainActivity.java [hunk 1]
```diff
@@ -125,17 +120,10 @@ public void onResume() {
             }
             DialogFragment dialog = new EnableNfcDialog();
             dialog.show(ft, "dialog");
-        } else if (totpListener != null) {
-            Intent intent = getIntent();
-            if (readOnResume && intent.getAction().equals(NfcAdapter.ACTION_NDEF_DISCOVERED)) {
```
Comment (petter-fidesmo) [inline src/main/com/yubico/yubioath/MainActivity.java]: @iDJH @Franzaine I think the removal of this is what discards the tag when the app is invoked via intent as described in #27 . My bad, will provide a fix.

**Pull request merged**